src/tsdb/prometheus.act

```diff
@@ -0,0 +1,147 @@
+            log.trace(response.body.decode(), None)
+
+            def _post_ready(timeout):
+                json_dict = { "name": "default" }
+                #http_client.post("/api/v1/services/prometheusdb/namespace/ready", json.encode(json_dict).encode(), lambda c, r: _on_ready_response(c, r, timeout))
+
+            def _on_ready_response(client: http.Client, response: http.Response, timeout: int):
+                log.trace("Namespace ready request:" + response.body.decode(), None)
+
+                if response.status != 200 and timeout > 0:
+                    after 1: _post_ready(timeout - 1)
+                else:
+                    await async on_response(self, response.status == 200)
+
+            _post_ready(20)
+
+        json_dict = { "type": "local", "namespaceName": "default", "retentionTime": "12h" }
+        #http_client.post("/api/v1/database/create", json.encode(json_dict).encode(), _on_response)
+
+    def send_metric(metric: Metric, on_response: action(Client, bool) -> None):
+
+        #mtags = {}
+        #for k, v in metric.tags.items():
+        #    mtags[k] = v
+
+        #mtags["__name__"] = metric.name
+        #json_dict = { "tags": mtags, "value": metric.value }
+
+        #json_dict["timestamp"] = str(metric.timestamp)
+
+        #def _on_response(client: http.Client, response: http.Response):
+        #    log.debug("HTTP response status: " + str(response.status), None)
+        #    log.trace(response.body.decode(), None)
+        #    await async on_response(self, response.status == 200)
+
+        #json_string = json.encode(json_dict)
+        #log.trace(json_string, None)
+
+        #if json_string is not None:
+        #    http_client.post("/api/v1/json/write", json_string.encode(), _on_response)
+        send_metrics([metric], on_response)
+
+    def send_metrics(metrics: list[Metric], on_response: action(Client, bool) -> None):
+        def _on_response(client: http.Client, response: http.Response):
+            log.debug("HTTP response status: " + str(response.status), None)
+            log.debug(response.body.decode(), None)
+
+        metric_data: list[(list[(str, str)], list[(float, int)])] = []
+        metric_dict: dict[str, list[Metric]] = {}
+        
+        for metric in metrics:
+            series_id = metric.name + str(metric.tags)
+
+            if series_id not in list(metric_dict.keys()):
+                 metric_dict[series_id] = []
+
+            metric_dict[series_id].append(metric)
+
+        for _, metric_series in metric_dict.items():
+            labels = [("__name__", metric_series[0].name)]
+            for name, value in metric_series[0].tags.items():
+                labels.append((name, value))
+
+            samples = []
+            for metric in metric_series:
+                samples.append((metric.value, metric.timestamp * 1000))
+
+            metric_data.append((labels, samples))
+
+        log.debug("metric_data: " + str(metric_data), None)
```

Comment:
The signature is `log.debug(str, ?dict[str, value])`

The idea with this logging interface is that you writes a fixed string about the event, which goes in the first argument. Then if you have variable data, you can add that in as the second argument and it's passed along as actually structured data. Right now that second arg has to be a dict keyed on str.

Passing along a dict, in particular if that is an existing dict, is very cheap since it's just a pointer. So if the log level is low so the message won't even be printed, this whole log.debug call turns pretty much into a noop and since we haven't done any other expensive operations, it's cheap. But if you first call `str()` on the dict, you will always perform that string formatting of the data regardless of whether the log message will later be logged or discarded based on the current log level. Please avoid that. Try and send 

Even constructing a new small dict is quite OK and relatively cheap (cheaper than recursive string formatting), something like:
```suggestion
        log.debug("send_metric metric_data", {"metric": metric_data})
```

Ideally we might want something even cheaper than this but it's quite hard without forcing more ritual on the programmer like:
```suggestion
        log.debug_enabled() and log.debug("metric_data: " + str(metric_data), None)
```
in which case we wouldn't actually run the second part at all and thus save even more CPU cycles